packages/server/src/middleware/modern.js

```diff
@@ -11,12 +13,33 @@ const isModernBrowser = (ua) => {
   })
 }
 
-export default function (req, res, next) {
-  const { socket = {}, headers } = req
-  if (socket.modernMode === undefined) {
-    const ua = headers && headers['user-agent']
-    socket.modernMode = isModernBrowser(ua)
+let detected = false
+
+const detectModernBuild = ({ options, resources }) => {
+  if (detected === false && ![false, 'client', 'server'].includes(options.modern)) {
+    detected = true
+    if (resources.modernManifest) {
+      options.modern = options.render.ssr ? 'server' : 'client'
+      consola.info(`Modern bundles are detected. Modern mode (${chalk.green.bold(options.modern)}) is enabled now.`)
+    } else {
+      options.modern = false
+    }
+  }
+}
+
+const detectModernBrowser = (req, options) => {
+  if (options.modern === 'server') {
+    const { socket = {}, headers } = req
+    if (socket.modernMode === undefined) {
+      const ua = headers && headers['user-agent']
+      socket.modernMode = isModernBrowser(ua)
```

Comment:
Why it is directly set on `socket` object?

Reply:
For reducing the matching times, we only match one time in a connection.

Reply:
Damn good idea 👍 It worth mentioning in the comments :D